File: old/Privault.Code/Models/Entities/AttachmentEntity.cs
using SQLite;

// ReSharper disable RedundantDefaultMemberInitializer
// ReSharper disable UnusedAutoPropertyAccessor.Global
// ReSharper disable AutoPropertyCanBeMadeGetOnly.Global
// ReSharper disable PropertyCanBeMadeInitOnly.Global

namespace Privault.Core.Models.Entities;

/// <summary>
/// Repräsentiert einen Dateianhang zu einem Tresoreintrag in der lokalen SQLite-Datenbank.
/// Der gesamte Inhalt wird verschlüsselt gespeichert, um die Privatsphäre zu gewährleisten.
/// <para>
/// <b>Sicherheitshinweise:</b>
/// <list type="bullet">
/// <item><b>Inhalt:</b> Der binäre Dateiinhalt ist mit dem AES-Schlüssel des zugehörigen Eintrags verschlüsselt.</item>
/// <item><b>Metadaten:</b> Metadaten wie Dateiname und MIME-Typ liegen als verschlüsselter JSON-Blob vor.</item>
/// </list>
/// </para>
/// </summary>
[Table("attachments")]
public class AttachmentEntity
{
    /// <summary>
    /// Die interne ID (Auto-Increment).
    /// Dient als Primärschlüssel für den Object-Relational Mapper (ORM).
    /// </summary>
    [PrimaryKey, AutoIncrement]
    [Column("id")]
    public int Id { get; set; }
    
    /// <summary>
    /// Die globale eindeutige ID des Anhangs (Universally Unique Identifier v4).
    /// </summary>
    [Indexed(Name = "uk_attachments_uuid", Unique = true)]
    [Column("uuid")]
    public string Uuid { get; set; } = string.Empty;

    /// <summary>
    /// Die interne ID des zugehörigen Eintrags.
    /// </summary>
    [Indexed(Name = "idx_attachments_entry_id")]
    [Column("entry_id")]
    public int EntryId { get; set; }
    
    /// <summary>
    /// Der AES-256-GCM verschlüsselte Metadaten-Container (Ciphertext + Nonce + Auth-Tag).
    /// Enthält das serialisierte JSON-Objekt der Klasse <see cref="Payloads.AttachmentMetaPayload"/>.
    /// </summary>
    [Column("encrypted_meta")]
    public string EncryptedMeta { get; set; } = string.Empty;

    /// <summary>
    /// Der AES-256-GCM verschlüsselte Binärdaten-Container (Ciphertext + Nonce + Auth-Tag).
    /// Enthält den binären Dateninhalt des Anhangs.
    /// </summary>
    [Column("encrypted_content")]
    public string EncryptedContent { get; set; } = string.Empty;
    
    /// <summary>
    /// <c>true</c>, wenn der Anhang synchronisiert wurde, sonst <c>false</c>
    /// </summary>
    [Indexed(Name = "idx_attachments_is_synced")]
    [Column("is_synced")]
    public bool IsSynced { get; set; } = false;
}

File: lib/models/entities/attachment_entity.dart
/// Repräsentiert einen Dateianhang zu einem Tresoreintrag in der lokalen SQLite-Datenbank.
/// Der gesamte Inhalt wird verschlüsselt gespeichert, um die Privatsphäre zu gewährleisten.
class AttachmentEntity {
  /// Die interne ID (Auto-Increment in der Datenbank).
  /// Nullable für neue Einträge, bevor sie in die Datenbank geschrieben werden.
  final int? id;

  /// Die globale eindeutige ID des Anhangs (Universally Unique Identifier v4).
  final String uuid;

  /// Die interne ID des zugehörigen Eintrags.
  final int entryId;

  /// Der AES-256-GCM verschlüsselte Metadaten-Container (Ciphertext + Nonce + Auth-Tag).
  /// Enthält das serialisierte JSON-Objekt der Klasse [AttachmentMetaPayload].
  final String encryptedMeta;

  /// Der AES-256-GCM verschlüsselte Binärdaten-Container (Ciphertext + Nonce + Auth-Tag).
  /// Enthält den binären Dateninhalt des Anhangs.
  final String encryptedContent;

  /// `true`, wenn der Anhang erfolgreich zum Server synchronisiert wurde, sonst `false`.
  final bool isSynced;

  /// Konstruktor
  AttachmentEntity({
    this.id,
    required this.uuid,
    required this.entryId,
    required this.encryptedMeta,
    required this.encryptedContent,
    this.isSynced = false,
  });

  /// Konvertiert eine [AttachmentEntity] in eine Map (z.B. für SQLite oder JSON).
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'uuid': uuid,
      'entry_id': entryId,
      'encrypted_meta': encryptedMeta,
      'encrypted_content': encryptedContent,
      'is_synced': isSynced ? 1 : 0,
    };
  }

  /// Erstellt ein [AttachmentEntity] Objekt aus einer Map.
  factory AttachmentEntity.fromMap(Map<String, dynamic> map) {
    return AttachmentEntity(
      id: map['id'] as int?,
      uuid: map['uuid'] as String,
      entryId: map['entry_id'] as int,
      encryptedMeta: map['encrypted_meta'] as String,
      encryptedContent: map['encrypted_content'] as String,
      isSynced: (map['is_synced'] as int) == 1,
    );
  }

  /// Erzeugt eine Kopie des Objekts mit modifizierten Eigenschaften.
  AttachmentEntity copyWith({int? id, String? uuid, int? entryId, String? encryptedMeta, String? encryptedContent, bool? isSynced}) {
    return AttachmentEntity(
      id: id ?? this.id,
      uuid: uuid ?? this.uuid,
      entryId: entryId ?? this.entryId,
      encryptedMeta: encryptedMeta ?? this.encryptedMeta,
      encryptedContent: encryptedContent ?? this.encryptedContent,
      isSynced: isSynced ?? this.isSynced,
    );
  }
}
